wal/wal.go

```diff
@@ -617,15 +617,23 @@ func (w *WAL) log(rec []byte, final bool) error {
 
 		// By definition when a record is split it means its size is bigger than
 		// the page boundary so the current page would be full and needs to be flushed.
-		// On contrary if we wrote a full record, we can fit more records of the batch
-		// into the page before flushing it.
-		if final || typ != recFull || w.page.full() {
+		// On contrary if we wrote a full record or last fragment of a record, we can
+		// fit more records of the batch into the page before flushing it.
+		if (typ != recFull && typ != recLast) || w.page.full() {
 			if err := w.flushPage(false); err != nil {
 				return err
 			}
 		}
 		rec = rec[l:]
 	}
+
+	// If it's the final record of the batch, flush the page.
+	if final {
```

Comment:
If the last record made the page full it will have already flushed from the logic above and then we will flush an empty page. That doesn't seem ideal.

Reply:
Fixed as well.